Not everyone has a cat; polyfill it

aschmolck opened the pull request:
A hard external dependency gone where bash gets the job done. It might often be
a few orders of magnitude slower for larger files, but computers are plenty fast
these days and users plenty patient. In fact for the intended use case dog may
be faster than cat:

```
~/repo/shab$ time dog example.shab >/dev/null
real	0m0.002s
user	0m0.001s
sys	0m0.000s

~/repo/shab$ time cat example.shab >/dev/null
real	0m0.004s
user	0m0.002s
sys	0m0.002s
```

Commit:
Not everyone has a cat; polyfill it

A hard external dependency gone where bash gets the job done. It might often be
a few orders of magnitude slower for larger files, but computers are plenty fast
these days and users plenty patient. In fact for the intended use case dog may
be faster than cat:

```
~/repo/shab$ time dog example.shab >/dev/null
real	0m0.002s
user	0m0.001s
sys	0m0.000s

~/repo/shab$ time cat example.shab >/dev/null
real	0m0.004s
user	0m0.002s
sys	0m0.002s
```

roberth commented:
Thank you! This seems to resolve long-standing bug #2. I have one or two concerns though. `dog` implements functionality that isn't used. It seems that it should work fine without the for loop.
My other concern is the dependency on device nodes. Instead of `/dev/null` we could pipe into the no-op builtin `:`. Perhaps `/dev/stdin` can be replaced by a file descriptor redirect or may be omitted entirely.

Commit:
fixup! Not everyone has a cat; polyfill it

- only emulate cat with up to one arg
- delete bogus `export -f`
- stop pandering to cat lovers
- embrace bashishm

aschmolck commented:
Concerning device nodes, good point. In the first case `&|:` or`>- 2>-` would do, but just dumping cat for dog is probably the altogether most elegant solution. I see no way to get rid of `/dev/stdin`  without code duplication or eval, so I opted for the latter.

I've also gotten rid of the loop, although I feel that `shab` accepting any number of files might not be a bad idea anyway – then it would act basically as templated cat.

aschmolck commented:
Actually, `if [[ "${1-}" = - ]]; then exec <"$1"; fi` plus subshell would probably also work instead of eval.

roberth commented:
I agree that we don't need `eval` for this. Perhaps another way we can avoid duplication is by extracting out a function that reads the lines from stdin and prints them to stdout.

zimbatm commented:
Some people are not dog persons and wouldn't like to ditch the cat for a dog.

If `dog()` was renamed to `cat()`, the user could decide if he wants to copy-and-paste just the `shab()` function in his container without the shim.

zimbatm commented:
merging that genius solution

zimbatm commented:
follow-up commit to make dog optional: abadd6b0d11252fc78dc766ab5d90c37da47c36c
